### PR TITLE
Fix zstd compression config handling

### PR DESCRIPTION
**Pull request by yh-sb**
The following issues are fixed:
* Allow zstd to decompress symlinks. "... is a symbolic link, ignoring" warning.
* Temp directory for compressed image should be created first. Otherwise, you will get "Directory nonexistent" error.
* Clarified "Wrong compression algorithm:" error message to show the list of supported values in "compression" option.

### CODE REVIEW

File: swugenerator/generator.py [hunk 1]
```diff
@@ -84,10 +84,11 @@ def process_entry(self, entry):
             if "compressed" in entry and not self.nocompress:
                 cmp = entry["compressed"]
                 if cmp not in ("zlib", "zstd"):
-                    logging.critical("Wrong compression algorithm: %s", cmp)
+                    logging.critical("Wrong compression algorithm: \"%s\". Supported only \"zlib\" or \"zstd\".", cmp)
```
Comment (obbardc) [inline swugenerator/generator.py:87]: now you have a duplicated list of supported compression algs, perhaps move that into its own variable

Comment (yh-sb) [reply swugenerator/generator.py]: I removed this from PR.

**Comment by obbardc** [comment]
LGTM.

**Comment by yh-sb** [comment]
So, will it be merged? 🙂

**Comment by sbabic** [comment]
> @sbabic, so, will it be merged? 🙂

I cannot merge because the commits are not formally ok. The two commits does not contain your Signed-off-by, and then I cannot suppose you agree with contributions' certificate in SWUpdate:

https://sbabic.github.io/swupdate/contributing.html#developer-s-certificate-of-origin-1-1

**Review by sbabic — COMMENTED** [review]
It is not clear to me what this should done. Change does not do what the commit message is saying.

The variable new_path is built with a path for a temporary directory (self.temp.name) concatenating the name of the filename found in sw-description and the suffix 'cmp'. It is not a directory path, it is the full path of the output of compression.
The change here just create a directory that is not used.. Am I wrong ?

**Comment by yh-sb** [comment]
`new_path` variable is passed as the last argument to the gzip or zstd:
https://github.com/sbabic/swugenerator/blob/57b84334803a1294336ae603d7cb85275f47a203/swugenerator/generator.py#L104
But gzip or zstd don't create this directory by themselves. So, we need to create this directory first.

Second commit is needed to pass `-f` option to zstd which allows to take symlink as an input path.

**Comment by sbabic** [comment]
> But **gzip** or **zstd** don't create output directory by themselves. So, we need to create this directory first.
> This fix is needed in case when no output directory is created at the moment when swugenerator is called.

Yes, but "new_path" is the full pathname of the result, that is the compressed file, and not of the directory.

The directory is created with the Temporary class:

                self.temp = TemporaryDirectory()

And the class creates the directory. new_path is just this temporary driectory + the filename with the "cmp" suffix, and not a directory. gzip and zstd write to stdout into "new_path", that is into a file.  So :+1: 

                         os.makedirs(os.path.dirname(new_path), exist_ok=True)

This seems wrong as new_path is thought to be a file not a directory. And creating the directory is superfluous because this is done by the Temporary class.

**Comment by yh-sb** [comment]
The error message I got without the fix:
`/bin/sh: 1: cannot create /tmp/tmp199uhqux/buildroot/output/images/rootfs.ext4.zlib: Directory nonexistent`

indicates that `new_path` contains **subdirectories** (`buildroot/output/images`) at the moment of calling **gzip** or **zstd**. 
These subdirectories weren't created by `self.temp = TemporaryDirectory()` at the beginning. Therefore, it is necessary to create these subdirectories if they do not exist.

Adding the line `os.makedirs(os.path.dirname(new_path), exist_ok=True)` ensures that all required directories are created, preventing the "*cannot create ... Directory nonexistent*" error.

**Comment by sbabic** [comment]
> The error message I got without the fix: `/bin/sh: 1: cannot create /tmp/tmp199uhqux/buildroot/output/images/rootfs.ext4.zlib: Directory nonexistent`

This raises another topic. Where are coming buildroot/output/images ? Have you put these in sw-description (it shouldn't be) ? Directories passed via --artifactory should be removed from the filename.
 
> indicates that `new_path` contains **subdirectories** (`buildroot/output/images`) at the moment of calling **gzip** or **zstd**. These subdirectories weren't created by `self.temp = TemporaryDirectory()` at the beginning. 

Sure - but the reason is that the generator shouldn't fill the filesystem. Just 'self.temp' should be used, and compressed files must be temporary stored in this directory.  So for other reason, a full path is taken, and this cause the issue you see, but let's say, the fix just avoid the error but it is not correct. The fix should avoid to use  /tmp/tmp199uhqux/buildroot/output/images/rootfs.ext4.zlib, and it should use  /tmp/tmp199uhqux/rootfs.ext4.zlib instead.
 
> Therefore, it is necessary to create these subdirectories if they do not exist.

No. The reason for the issue is not here, and this is a work-around.

**Comment by yh-sb** [comment]
This path with subdirectories comes from
`--artifactory ./buildroot/output/images` or `--swu-file ./buildroot/output/images/rpi-board.swu` options.

**Comment by sbabic** [comment]
> This path with subdirectories comes from `--artifactory ./buildroot/output/images` or `--swu-file ./buildroot/output/images/rpi-board.swu` options.

I have tried to reproduce your issue. I put a "rootfs" into a "test" directory, and I ran with :+1: 
       
                         -o ./test/SWUfile-test2.swu -a ./deploy,./test 

and deploy is a fake, it does not exist. Just to check if it can be an issue. I have a rootfs just in "test" and according to the issue, I should run into problems because it should try to compress a file like /tmp/<generated dir>/test/<my rootfs>.ext4.

I ran the debugger to check the vars, and before running the compressor (zlib), vars do not have "test". So in my test any path due to --artifactory or --swu-file are removed, I do not know why they are still in your environment. I do not think they are due because you applied to an older commit, last commits have nothing to do with this. Everything seem working.

**Comment by yh-sb** [comment]
Ok, I don't have setup to reproduce this. I will close this PR.

**Pull request closed without merging**